### ISSUE TITLE
Vue router lazy components not working

### ISSUE THREAD

**Issue by sentiasa** [spatie/laravel-server-side-rendering#49]
I have explained the issue in detail here:
https://github.com/JeffreyWay/laravel-mix/issues/2245

To recap, this works:
```
import Home from '../view/Home';
export const routes = [{ path: '/', name: "Home", component: Home }]
```
...however these don't
```
const Home = resolve => require(['../views/Home.vue'], m => resolve(m.default));
export const routes = [{ path: '/', name: "Home", component: Home }]
```
> `<div id="app" data-server-rendered="true"><!----></div>`
------

```
export const routes = [{ path: '/', name: "Home", component: () => import('../views/Home.vue)' }]
```
> `<div id="app" data-server-rendered="true"><!----></div>`
-----

For the `() => import('../views/Home.vue)`, client.js works, only server.js doesn't work.

I suspect it's because server.js doesn't wait for the dynamic routes to be loaded, however I couldn't find any solution. Thanks in advance

**Comment by sentiasa** [comment]
I think I have an idea why this is happening:
```
export const routes = [{
     path: '/', 
     name: "Home", 
     component: () => import('../views/Home')
}]
```
with this code, I get an error:
> Error: Cannot find module './js/chunks/server/0.js?id=c3384f174123f0848451'

> The command "/usr/bin/node /home/vagrant/Code/project/storage/app/ssr/717358e60bfd52035a1e58256cdfbba0.js" failed. Exit Code: 1(General error) Working directory: /home/vagrant/Code/project/public Output: ================ Error Output: ================ internal/modules/cjs/loader.js:628 throw err; ^ Error: Cannot find module './js/chunks/server/0.js?id=c3384f174123f0848451'

Look at the paths:
In my compiled file (which is at `public/js`) I have this line:
```
var chunk = require("./js/chunks/server/" + ({}[chunkId]||chunkId) + ".js?id=" + {"0":"c3384f174123f0848451"}[chunkId] + "");
```
That seems like a relative path. However the file is actually running in what I specify in the `config/ssr.php` - `'temp_path' => storage_path('app/ssr')` - so it cannot find the path.

However, even I change the `temp_path` to `public_path()` so that it can find the chunk from  `./js/chunks/server/` (which is `public/js/chunks/server/0.js`), it still throws the same error. Even though the SSR's temp_path is different.

> The command "/usr/bin/node /home/vagrant/Code/project/public/3560d8d101faa4bdef316054b14873cc.js" failed. Exit Code: 1(General error) Working directory: /home/vagrant/Code/project/public Output: ================ Error Output: ================ internal/modules/cjs/loader.js:628 throw err; ^ Error: Cannot find module './js/chunks/server/0.js?id=c3384f174123f0848451'

@sebastiandedeyne Can you please take a look?

**Comment by sentiasa** [comment]
If anyone wonders - this is [how I achieved it](https://stackoverflow.com/questions/58216963/lazy-loaded-vue-router-components-dont-work-with-ssr/58297349#58297349):

I learn about separating bundles from [charlesBochet's comment](https://github.com/JeffreyWay/laravel-mix/issues/989#issuecomment-440756310), however instead of 2x `webpack.mix.js` files I used one.

- package.json
```
"scripts": {
    "dev-all": "concurrently \"npm --section=server run dev\" \"npm --section=client run dev\"  --kill-others-on-fail",
    "dev-client": "npm --section=client run dev",
    "dev-server": "npm --section=server run dev"
     ...
}
```

- webpack.mix.js

```
if (process.env.npm_config_section === 'server') {
    mix.js('resources/js/app-server.js', 'public/js')
        .webpackConfig({
            target: 'node',
  
            // Prevent code-split for server-build
            plugins: [
                new webpack.optimize.LimitChunkCountPlugin({
                    maxChunks: 1,
                })
            ],
        })
        .mergeManifest()
        .version();

} else if (process.env.npm_config_section === 'client') {
    mix.js('resources/js/app-client.js', 'public/js')
        .webpackConfig({
            target: 'web',
            output: {
                chunkFilename: 'js/chunks/[name].js?id=[chunkhash]',
                publicPath: '/',
            },
        })
        .mergeManifest()
        .version();

    // Only build css with the client build, server build only needs
    // the html and not the css
    mix.sass('resources/sass/app.scss', 'public/css')
} else {
    console.log(
        '\x1b[41m%s\x1b[0m',
        'Provide correct --section argument to build command: server, client'
    );
    throw new Error('Provide correct --section argument to build command!')
}
```

- app-server.js - should wait for the router to be ready
```
new Promise((resolve, reject) => {
    router.push(context.url);
    router.onReady(() => {
        resolve(app);
    }, reject);
})
    .then(app => {
        renderVueComponentToString(app, (err, res) => {
            if (err) throw new Error(err);
            dispatch(res);
        });
    });
```

- and finally router file code-splitting works for app-client.js
```
export const routes = [
{
    path: '/',
    name: "Home",
    component: () => import('../views/Home.vue')
},
```

**Comment by clonesia** [comment]
why I got this error?

> vendor.js:26457 [Vue warn]: Failed to mount component: template or render function not defined.
> 
> found in
> 
> ---> <Anonymous>
>        <App> at resources/js/src/App.vue
>          <Root>

**Comment by atodd** [comment]
Just wrestled with this all day.  In addition to the solutions above, separate webpack or mix configs for client and server, and waiting on vue-router to be ready.  I had to change the output path in webpack to match the `storage/app/ssr` directory being used for the temp file.  The problem was the temp file was stored in `storage/app/ssr` and my async components were in `public/dist`.

Used the storage path helper in my blade template
`{!! ssr(storage_path('app/ssr/js/app-server.js'))->render() !!}`

Changed Webpack config to
`path: path.resolve(__dirname, 'storage/app/ssr'),`

It's all working now.  Hope this helps anyone else who was stuck on this like me.
